src/coreclr/src/vm/genericdict.cpp

```diff
@@ -640,6 +769,152 @@ Dictionary::Restore(
+        PRECONDITION(pMT->GetClass()->GetDictionaryLayout()->GetMaxSlots() > 0);
+    }
+    CONTRACTL_END;
+
+    TADDR* pDictionarySlots = (TADDR*)pMT->GetPerInstInfo()[pMT->GetGenericsDictInfo()->m_wNumDicts - 1].GetValue();
+    TADDR* pSizeSlot = pDictionarySlots + pMT->GetGenericsDictInfo()->m_wNumTyPars;
+    return *(DWORD*)pSizeSlot;
+}
+
+DWORD Dictionary::GetDictionarySlotsSizeForMethod(MethodDesc* pMD)
+{
+    CONTRACTL
+    {
+        PRECONDITION(pMD->AsInstantiatedMethodDesc()->IMD_GetMethodDictionary() != NULL);
+    }
+    CONTRACTL_END;
+
+    InstantiatedMethodDesc* pIMD = pMD->AsInstantiatedMethodDesc();
+    TADDR* pDictionarySlots = (TADDR*)pIMD->IMD_GetMethodDictionary();
+    TADDR* pSizeSlot = pDictionarySlots + pIMD->GetNumGenericMethodArgs();
+    return *(DWORD*)pSizeSlot;
+}
+
+Dictionary* Dictionary::GetMethodDictionaryWithSizeCheck(MethodDesc* pMD, ULONG slotIndex)
+{
+    CONTRACT(Dictionary*)
+    {
+        THROWS;
+        GC_TRIGGERS;
+        POSTCONDITION(CheckPointer(RETVAL));
+    }
+    CONTRACT_END;
+
+    Dictionary* pDictionary = pMD->GetMethodDictionary();
+
+#if !defined(CROSSGEN_COMPILE)
+    DWORD currentDictionarySize = GetDictionarySlotsSizeForMethod(pMD);
+
+    if (currentDictionarySize <= (slotIndex * sizeof(DictionaryEntry)))
+    {
+        // Only expand the dictionary if the current slot we're trying to use is beyond the size of the dictionary
+
+        // Take lock and check for size again, just in case another thread already resized the dictionary
+        CrstHolder ch(&SystemDomain::SystemModule()->m_DictionaryCrst);
+
+        pDictionary = pMD->GetMethodDictionary();
+        currentDictionarySize = GetDictionarySlotsSizeForMethod(pMD);
+
+        if (currentDictionarySize <= (slotIndex * sizeof(DictionaryEntry)))
+        {
+            DictionaryLayout* pDictLayout = pMD->GetDictionaryLayout();
+            InstantiatedMethodDesc* pIMD = pMD->AsInstantiatedMethodDesc();
+            _ASSERTE(pDictLayout != NULL && pDictLayout->GetMaxSlots() > 0);
+
+            DWORD expectedDictionarySize = DictionaryLayout::GetDictionarySizeFromLayout(pMD->GetNumGenericMethodArgs(), pDictLayout);
+            _ASSERT(currentDictionarySize < expectedDictionarySize);
+
+            pDictionary = (Dictionary*)(void*)pIMD->GetLoaderAllocator()->GetHighFrequencyHeap()->AllocMem(S_SIZE_T(expectedDictionarySize));
+
+            // Copy old dictionary entry contents
+            DictionaryEntry* pOldEntriesPtr = (DictionaryEntry*)pIMD->m_pPerInstInfo.GetValue();
+            DictionaryEntry* pNewEntriesPtr = (DictionaryEntry*)pDictionary;
+            for (int i = 0; i < currentDictionarySize / sizeof(DictionaryEntry); i++, pOldEntriesPtr++, pNewEntriesPtr++)
+            {
+                DictionaryEntry pSlotValue = VolatileLoadWithoutBarrier(pOldEntriesPtr);
+                VolatileStoreWithoutBarrier(pNewEntriesPtr, pSlotValue);
```

Comment:
Nit: This can be regular store. No other threads has access to the new entries at this point.